lmfdb/classical_modular_forms/test_cmf.py

```diff
@@ -937,4 +909,53 @@ def test_is_self_dual(self):
         for elt in ['52.1.j.a', '57.1.h.a', '111.1.h.a']:
             assert elt in page.get_data(as_text=True)
 
+    def test_hecke_charpolys(self):
+        """Test that the Hecke charpolys are correct.
+
+        Some expected Hecke charpolys are stored in the dict test_data,
+        which are then checked to be in the relevant page. These examples
+        have been chosen to be readily verifiable from the displayed
+        Fourier coefficients of each respective homepage."""
+
+        test_data = {
+                    '11/2/a/a' : {2: '\( 2 + T \)',
+                                 17: '\( 2 + T \)',
+                                 29: '\( T \)'},
+
+                    '10/3/c/a' :  {5: r'\( 25 + T^{2} \)',
+                                  11: r'\( ( 8 + T )^{2} \)',
+                                  97: r'\( 7938 + 126 T + T^{2} \)'},
+
+                    '/294/5/b/f' : {2: r'\( ( 8 + T^{2} )^{5} \)',
+                                    # The following test checks that monomials do not have superfluous parentheses
+                                    7: r'\( T^{10} \)'},
+
+                    # The following is a newspace with 5 newforms, hence the more verbose polynomials.
+                    # These newspace examples have historically also been prone to superfluous parentheses
+                    # (e.g. LMFDB Issue #4117)
+
+                    '255/1/h/' : {2: r'(\( 1 + T \))(\( 1 + T \))(\( -1 + T \))(\( -1 + T \))(\( T^{2} \))',
+                                  3: r'(\( 1 + T \))(\( -1 + T \))(\( 1 + T \))(\( -1 + T \))(\( 1 + T^{2} \))',
+                                  5: r'(\( 1 + T \))(\( -1 + T \))(\( -1 + T \))(\( 1 + T \))(\( 1 + T^{2} \))'},
+
+                    # The following is a newspace with 9 newforms, but only 7 have been computed exactly.
+                    # This also checks correct parentheses
+
+                    '4866/2/a/' : {2: r'\( ( 1 + T )^{2} \)\( ( -1 + T )^{11} \)\( ( 1 + T )^{13} \)\( ( 1 + T )^{14} \)\( ( -1 + T )^{14} \)\( ( 1 + T )^{17} \)\( ( -1 + T )^{20} \)',
+                                   3: r'\( ( -1 + T )^{2} \)\( ( -1 + T )^{11} \)\( ( 1 + T )^{13} \)\( ( -1 + T )^{14} \)\( ( 1 + T )^{14} \)\( ( -1 + T )^{17} \)\( ( 1 + T )^{20} \)'},
+
+                    # The last one is the original issue highlighted in #4117
+
+                    '81/2/c/' : {# We want to put brackets around each monomial power for newspaces, to make it clear that we are taking a product over multiple distinct newforms
+                                 3: r'(\( T^{2} \))(\( T^{4} \))',
```

Comment:
I assumed that in this case, we *do* want parentheses around the $T^2$ and $T^4$, to make it clear that the hecke characteristic polynomial is a product coming from each of the newforms, but let me know if it should be otherwise

Reply:
I think we should treat all factored polynomials the same and not try to show any extra structure by adding parentheses.

Reply:
The parentheses around T^2 and T^4 were arising because this is a newspace with two newform subspaces, and the present code was concatenating the factorisations rather than multiplying them and then factorising (as you noted below with 5550.2.a). As discussed below, this will be addressed by not showing these polynomials for newspaces.